### ISSUE TITLE
fastapi return 500 instead of 404 if url has a @

### ISSUE THREAD

**Issue by sileht** [DataDog/dd-trace-py#4932]
### Summary of problem

Got an HTTP 500 instead of an expected 404 error on a fastapi application.

### Which version of dd-trace-py are you using?

1.7.1

### Which version of pip are you using?

22.3.1

### Which libraries and their versions are you using?

<details>
  <summary>`pip freeze`</summary>
```
aiofiles==22.1.0
alabaster==0.7.12
anyio==3.6.2
anys==0.2.1
asgi-lifespan==1.0.1
async-timeout==4.0.2
asyncpg==0.27.0
attrs==21.4.0
Authlib==1.2.0
Babel==2.11.0
beautifulsoup4==4.11.1
black==22.12.0
boltons==21.0.0
bracex==2.3.post1
bytecode==0.14.0
cattrs==22.2.0
certifi==2022.12.7
cffi==1.15.1
charset-normalizer==2.1.1
click==8.1.3
click-option-group==0.5.5
colorama==0.4.6
commonmark==0.9.1
coverage==7.0.2
cryptography==39.0.0
daiquiri==3.2.1
datadog==0.44.0
ddsketch==2.0.4
ddtrace==1.6.4
deepdiff==6.2.2
defusedxml==0.7.1
Deprecated==1.2.13
docstring-to-markdown==0.11
docutils==0.19
dulwich==0.20.50
envier==0.4.0
execnet==1.9.0
face==22.0.0
fastapi==0.88.0
filelock==3.9.0
first==2.0.2
flake8==5.0.4
flake8-black==0.3.6
flake8-blind-except==0.2.1
flake8-bugbear==22.12.6
flake8-builtins==2.1.0
flake8-comprehensions==3.10.1
flake8-docstrings==1.6.0
flake8-logging-format==0.8.1
flake8-new-union-types==0.4.1
flake8-pep585==0.1.6
flake8-rst-docstrings==0.2.7
flake8-use-fstring==1.4
freezegun==1.2.2
glom==22.1.0
graphql-utils==0.4
greenlet==2.0.1
gunicorn==20.1.0
h11==0.14.0
h2==4.1.0
hiredis==2.1.0
honcho==1.1.0
hpack==4.0.0
httpcore==0.16.3
httptools==0.5.0
httpx==0.23.2
hyperframe==6.0.1
idna==3.4
imagesize==1.4.1
imia==0.5.3
iniconfig==1.1.1
isort==5.11.4
itsdangerous==2.1.2
jedi==0.18.2
jedi-language-server==0.40.0
Jinja2==3.1.2
jsonschema==4.17.3
lsprotocol==2022.0.0a9
markdownify==0.11.6
MarkupSafe==2.1.1
mccabe==0.7.0
msgpack==1.0.4
multidict==6.0.4
mypy==0.990
mypy-extensions==0.4.3
ordered-set==4.1.0
packaging==21.3
parso==0.8.3
passlib==1.7.4
pastel==0.2.1
pathspec==0.10.3
pbr==5.11.0
peewee==3.15.4
platformdirs==2.6.2
pluggy==1.0.0
poethepoet==0.16.5
protobuf==4.21.12
py==1.11.0
pycodestyle==2.9.1
pycparser==2.21
pydantic==1.10.4
pydocstyle==6.2.2
pyenchant==3.2.2
pyflakes==2.5.0
pygls==1.0.0
Pygments==2.13.0
PyJWT==2.6.0
pyparsing==3.0.9
pyrsistent==0.19.3
pytest==7.2.0
pytest-asyncio==0.20.3
pytest-cov==4.0.0
pytest-forked==1.4.0
pytest-github-actions-annotate-failures==0.1.8
pytest-httpserver==1.0.6
pytest-split==0.8.0
pytest-timeout==2.1.0
pytest-xdist==2.5.0
python-dateutil==2.8.2
python-dotenv==0.21.0
python-json-logger==2.0.4
python-lsp-jsonrpc==1.0.0
python-multipart==0.0.5
pytz==2022.7
PyYAML==6.0
redis==4.4.0
reno==3.5.0
requests==2.28.1
respx==0.20.1
restructuredtext-lint==1.4.0
rfc3986==1.5.0
rich==12.6.0
ruamel.yaml==0.17.21
semgrep==0.118.0
sentry-sdk==1.9.10
setproctitle==1.2.3
six==1.16.0
sniffio==1.3.0
snowballstemmer==2.2.0
soupsieve==2.3.2.post1
Sphinx==5.3.0
sphinxcontrib-applehelp==1.0.2
sphinxcontrib-devhelp==1.0.2
sphinxcontrib-htmlhelp==2.0.0
sphinxcontrib-jsmath==1.0.1
sphinxcontrib-qthelp==1.0.3
sphinxcontrib-serializinghtml==1.1.5
sphinxcontrib-spelling==7.7.0
SQLAlchemy==1.4.45
SQLAlchemy-Utils==0.39.0
sqlalchemy2-stubs==0.0.2a31
starlette==0.22.0
starsessions==2.1.1
tenacity==8.1.0
tokenize-rt==5.0.0
toml==0.10.2
tomli==2.0.1
tqdm==4.64.1
tryceratops==1.1.0
typeguard==2.13.3
types-first==2.0.5
types-freezegun==1.1.10
types-pkg-resources==0.1.3
types-PyYAML==6.0.12.2
types-redis==4.3.21.7
typing_extensions==4.4.0
ujson==5.6.0
urllib3==1.26.13
uvicorn==0.20.0
uvloop==0.17.0
vcrpy==4.2.1
voluptuous==0.13.1
watchfiles==0.18.1
wcmatch==8.4.1
websockets==10.4
Werkzeug==2.2.2
wrapt==1.14.1
xmltodict==0.13.0
yamllint==1.28.0
yarl==1.8.2
yesqa==1.4.0
```
</details>

### How can we reproduce your problem?

Making a request with a "@" in the path on a fastapi application that should return a 404 because the endpoint didn't exist like:

https://www.example.com/12345@

### What is the result that you get?

A HTTP 500 with the following backtrace:

```
  File "starlette/middleware/httpsredirect.py", line 19, in __call__
    await self.app(scope, receive, send)
  File "ddtrace/contrib/asgi/middleware.py", line 167, in __call__
    trace_utils.set_http_meta(
  File "ddtrace/contrib/trace_utils.py", line 444, in set_http_meta
    url = _sanitized_url(url)
  File "ddtrace/contrib/trace_utils.py", line 292, in _sanitized_url
    netloc = netloc[netloc.index("@") + 1 :]
ValueError: substring not found
```

### What is the result that you expected?

No error 500, but fastapi returning its 404

**Comment by majorgreys** [comment]
#4914 fixes this which will be released with 1.7.2:

https://github.com/DataDog/dd-trace-py/releases/tag/v1.7.2

This release was published just now so wheels are not yet uploaded to PyPI.

**Comment by brettlangdon** [comment]
1.7.2 and 1.6.6 have been released with a fix for this. 0.61.5 is still in progress.

@sileht thank you for reporting this!
